src/controller/subtitle-stream-controller.js

```diff
@@ -28,186 +36,208 @@ class SubtitleStreamController extends TaskLoop {
     if (!this.tracks || this.currentTrackId === -1) {
+      this.clearInterval();
       return;
     }
 
-    // Check if track was already loaded and if so make sure we finish
-    // downloading its frags, if not all have been downloaded yet
+    // Check if track has the necessary details to load fragments
     const currentTrack = this.tracks[this.currentTrackId];
     if (currentTrack && currentTrack.details) {
-      this.tick();
+      this.setInterval(TICK_INTERVAL);
     }
   }
 
   // Got a new set of subtitle fragments.
-  onSubtitleTrackLoaded () {
-    this.tick();
+  onSubtitleTrackLoaded (data) {
+    const { id, details } = data;
+
+    if (!this.tracks) {
+      logger.warn('Can not update subtitle details, no tracks found');
+      return;
+    }
+
+    if (this.tracks[id]) {
+      logger.log('Updating subtitle track details');
+      this.tracks[id].details = details;
+    }
+
+    this.setInterval(TICK_INTERVAL);
   }
 
   onKeyLoaded () {
     if (this.state === State.KEY_LOADING) {
       this.state = State.IDLE;
-      this.tick();
     }
   }
 
   onFragLoaded (data) {
-    let fragCurrent = this.fragCurrent,
-      decryptData = data.frag.decryptdata;
-    let fragLoaded = data.frag,
-      hls = this.hls;
+    const fragCurrent = this.fragCurrent;
+    const decryptData = data.frag.decryptdata;
+    const fragLoaded = data.frag;
+    const hls = this.hls;
+
     if (this.state === State.FRAG_LOADING &&
         fragCurrent &&
         data.frag.type === 'subtitle' &&
         fragCurrent.sn === data.frag.sn) {
       // check to see if the payload needs to be decrypted
-      if ((data.payload.byteLength > 0) && (decryptData != null) && (decryptData.key != null) && (decryptData.method === 'AES-128')) {
-        let startTime;
-        try {
-          startTime = performance.now();
-        } catch (error) {
-          startTime = Date.now();
-        }
+      if ((data.payload.byteLength > 0) &&
+          (decryptData != null) && (decryptData.key != null) &&
+          (decryptData.method === 'AES-128')) {
+        let startTime = performance.now();
+
         // decrypt the subtitles
         this.decrypter.decrypt(data.payload, decryptData.key.buffer, decryptData.iv.buffer, function (decryptedData) {
-          let endTime;
-          try {
-            endTime = performance.now();
-          } catch (error) {
-            endTime = Date.now();
-          }
+          let endTime = performance.now();
+
           hls.trigger(Event.FRAG_DECRYPTED, { frag: fragLoaded, payload: decryptedData, stats: { tstart: startTime, tdecrypt: endTime } });
         });
       }
     }
   }
+
+  doTick () {
+    if (!this.media) {
+      this.state = State.IDLE;
+      return;
+    }
+
+    switch (this.state) {
+    case State.IDLE:
+      const tracks = this.tracks;
+      const trackId = this.currentTrackId;
+
+      if (!tracks || trackId < 0 || trackId >= tracks.length || !tracks[trackId].details) {
```

Comment:
```suggestion
       if (!tracks || !tracks[trackId] || !tracks[trackId].details) { 
```

I think this is a bit more concise, and guards a bit more effectively